consensus/polybft/validator_set.go

```diff
@@ -0,0 +1,55 @@
+package polybft
+
+import (
+	"github.com/0xPolygon/pbft-consensus"
+	"github.com/0xPolygon/polygon-edge/types"
+)
+
+// ValidatorSet is a wrapper interface around pbft.ValidatorSet and it holds current validator set
+type ValidatorSet interface {
+	pbft.ValidatorSet
+
+	Accounts() AccountSet
+}
+
+type validatorSet struct {
+	// last proposer of a block
+	last types.Address
+
+	// current list of validators (slice of (Address, BlsPublicKey) pairs)
+	validators AccountSet
+}
+
+func newValidatorSet(lastProposer types.Address, validators AccountSet) *validatorSet {
+	return &validatorSet{validators: validators, last: lastProposer}
+}
+
+func (v validatorSet) Accounts() AccountSet {
+	return v.validators
+}
+
+func (v validatorSet) CalcProposer(round uint64) pbft.NodeID {
+	var seed uint64
+	if v.last == (types.Address{}) {
+		seed = round
+	} else {
+		offset := 0
+		if indx := v.validators.Index(v.last); indx != -1 {
+			offset = indx
+		}
+
+		seed = uint64(offset) + round + 1
+	}
+
+	pick := seed % uint64(v.validators.Len())
+
+	return pbft.NodeID(v.validators[pick].Address.String())
+}
```

Comment:
This method is not very truthful, if validators are simply appended to the end of the validator set. Validators that are added to the end can get the chance to propose the block before some other nodes which were next in line.
Example:
* Validators are A, B, C
* Block 0 is proposed by A, block 1 by B, block 2 by C
* When C is the proposer, a new validator is added to the end (D)
* The proposer for block 3 should've been A, but from this method it's D (which is unfair to A)
* Validator D should've been the proposer in the next round of proposals, not the current one

Reply:
> This method is not very truthful

As with the validator uptime comment. I think it is hard to come up with a silver bullet strategy for every piece of consensus logic. This in particular is based on the Tendermint proposer calculation which improves some of the security problems in Round Robin for public networks.